### ISSUE TITLE
MPO doesn't allow complex coefficients when QN-conserving?

### ISSUE THREAD

**Issue by sujay-kazi** [ITensor/ITensors.jl#318]
I have a question regarding a particular error I am getting. Basically, I want to implement an operator that looks like
Op = -ix*(sum from j=1 to N-2)[(S-)j * (Sz)(j+1) * (S+)(j+2) - (S+)j * (Sz)(j+1) * (S-)(j+2)]
as a pseudo-momentum operator (here N is an integer and x is a positive real number). In general, I want to able to compute <psi|Op^2|psi> for a state |psi> expressed as an MPS, which is why I am creating Op as an MPO. However, when I run the following code:

```
using ITensors
N = 10
x = 1
sites = siteinds("S=1/2", N, conserve_sz=true)
ampo_Op = AutoMPO()
for j = 1:N-2
  add!(ampo_Op,x,"Sx",j,"Sz",j+1,"Sy",j+2)
  add!(ampo_Op,-x,"Sy",j,"Sz",j+1,"Sx",j+2)
end
Op = MPO(ampo_Op, sites)
```

I get the following error:

```
ERROR: InexactError: Float64(0.0 - 1.0im)
Stacktrace:
 [1] Real at .\complex.jl:37 [inlined]       
 [2] convert at .\number.jl:7 [inlined]      
 [3] setindex! at .\array.jl:782 [inlined]   
 [4] setindex! at .\subarray.jl:295 [inlined]
 [5] setindex! at C:\Users\user\.julia\packages\ITensors\MgdBe\NDTensors\src\tensorstorage.jl:30 [inlined]
 [6] setindex! at C:\Users\user\.julia\packages\ITensors\MgdBe\NDTensors\src\dense.jl:226 [inlined]
 [7] copyto!(::IndexLinear, ::ITensors.NDTensors.Tensor{Float64,2,ITensors.NDTensors.Dense{Float64,SubArray{Float64,1,Array{Float64,1},Tuple{UnitRange{Int64}},true}},Tuple{Int64,Int64}}, ::IndexLinear, ::Array{Complex{Float64},2}) at .\abstractarray.jl:807
 [8] copyto! at .\abstractarray.jl:799 [inlined]
 [9] copyto! at .\broadcast.jl:863 [inlined]
 [10] materialize!(::ITensors.NDTensors.Tensor{Float64,2,ITensors.NDTensors.Dense{Float64,SubArray{Float64,1,Array{Float64,1},Tuple{UnitRange{Int64}},true}},Tuple{Int64,Int64}}, ::Base.Broadcast.Broadcasted{Base.Broadcast.DefaultArrayStyle{2},Nothing,typeof(identity),Tuple{Array{Complex{Float64},2}}}) at .\broadcast.jl:822
 [11] #qn_svdMPO#300(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::typeof(ITensors.qn_svdMPO), ::AutoMPO, ::Array{Index{Array{Pair{QN,Int64},1}},1}) at C:\Users\user\.julia\packages\ITensors\MgdBe\src\physics\autompo.jl:547
 [12] qn_svdMPO at C:\Users\user\.julia\packages\ITensors\MgdBe\src\physics\autompo.jl:362 [inlined]
 [13] #MPO#330 at C:\Users\user\.julia\packages\ITensors\MgdBe\src\physics\autompo.jl:734 [inlined]
 [14] MPO(::AutoMPO, ::Array{Index{Array{Pair{QN,Int64},1}},1}) at C:\Users\user\.julia\packages\ITensors\MgdBe\src\physics\autompo.jl:732
 [15] top-level scope at none:0
```

This seems to be happening because the MPO doesn't like non-real coefficients in the terms added to the AutoMPO object. I believe this because, when I write a similar-looking operator above as shown below, I no longer get an error. Notice that now the coefficients are real as opposed to imaginary:

```
using ITensors
N = 10
x = 1
sites = siteinds("S=1/2", N, conserve_sz=true)
ampo_Op = AutoMPO()
for j = 1:N-2
  add!(ampo_Op,x,"S-",j,"Sz",j+1,"S+",j+2)
  add!(ampo_Op,x,"S+",j,"Sz",j+1,"S-",j+2)
end
Op = MPO(ampo_Op, sites)
```

Is there a reason for this particular restriction? I am assuming there is, but it would be nice to have some clarity on this. I also wonder if this is related to #317, because there I have written the pseudo-momentum operator in a different way (using Sx and Sy instead of S+ and S-), and there I get a different error. In this particular case, the error also only occurs if you set either conserve_qns or conserve_sz to true, so I think the errors might be related.

**Comment by emstoudenmire** [comment]
Thanks - yes there was a bug here. Just pushed a fix for it in PR #319

**Comment by emstoudenmire** [comment]
Just to note: there's no restriction about using real versus complex coefficients in AutoMPO. It was a bug that occurred for complex coefficients with QN-conserving ITensors. Thanks again for pointing this out!
